### PR TITLE
SONARJAVA-3513 Improve S5810 to support static and test methods with return values

### PR DESCRIPTION
**Pull request by alban-auzeill**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/tests/AbstractJUnit5NotCompliantModifierChecker.java [hunk 1]
```diff
@@ -24,45 +24,56 @@
...
-      .filter(modifier -> isNotCompliant(modifier.modifier()))
+      .filter(modifier -> isNotCompliantModifier(modifier.modifier(), isMethod))
       .findFirst()
       .ifPresent(modifier -> reportIssue(modifier, "Remove this '" + modifier.keyword().text() + "' modifier."));
   }
 
+  private void raiseIssueOnNotCompliantReturnType(MethodTree methodTree) {
+    // Return type of METHOD is never null (unlike CONSTRUCTOR)
+    TypeTree returnType = methodTree.returnType();
+    if (isNotCompliantReturnType(returnType, methodTree.symbol())) {
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/tests/AbstractJUnit5NotCompliantModifierChecker.java:74]: I understand that you wanted to follow the same pattern as `raiseIssueOnNotCompliantModifiers`, but this code can report an issue only in one of the children, does it make sense to keep it in the common part?

File: java-checks/src/test/java/org/sonar/java/checks/tests/JUnit5DefaultPackageClassAndMethodCheckTest.java [hunk 2]
```diff
@@ -34,7 +34,10 @@ void test() {
       .onFile(testSourcePath)
       .withCheck(new JUnit5DefaultPackageClassAndMethodCheck())
       .verifyIssues();
+  }
 
+  @Test
+  void test_without_semantic() {
```
[the remark this answers is not in the document]
Comment (alban-auzeill) [reply java-checks/src/test/java/org/sonar/java/checks/tests/JUnit5DefaultPackageClassAndMethodCheckTest.java:40]: Nope. Even if we remove the `hasSemantic()` check, we still want to check how the rule implementation behaves with a lot of unknown symbols. So `.withoutSemantic()` is still useful. #UnknownSymbolsMatter

File: java-checks/src/test/java/org/sonar/java/checks/tests/JUnit5SilentlyIgnoreClassAndMethodCheckTest.java [hunk 3]
```diff
@@ -22,24 +22,36 @@
...
     JavaCheckVerifier.newVerifier()
-      .onFile(testSourcePath)
-      .withCheck(new JUnit5PrivateClassAndMethodCheck())
+      .onFile(testSourcesPath(SOURCE_PATH))
+      .withCheck(new JUnit5SilentlyIgnoreClassAndMethodCheck())
       .verifyIssues();
+  }
 
+  @Test
+  void test_without_semantic() {
```
[the remark this answers is not in the document]
Comment (alban-auzeill) [reply java-checks/src/test/java/org/sonar/java/checks/tests/JUnit5SilentlyIgnoreClassAndMethodCheckTest.java:41]: See above #UnknownSymbolsMatter comment.

**Review by quentin-jaquier-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**